src/quic/session.cc

```diff
@@ -0,0 +1,3506 @@
+      if (!packet) {
+        session_->last_error_ = QuicError::ForNgtcp2Error(NGTCP2_ERR_INTERNAL);
+        return session_->CloseSilently();
+      }
+      pos = ngtcp2_vec(*packet).base;
+    }
+
+    ssize_t nwrite = WriteVStream(&path, pos, &ndatalen, stream_data);
+
+    DEBUG_ARGS(session_, "Total amount written to packet: %" PRId64, nwrite);
+    DEBUG_ARGS(
+        session_, "Stream data amount written to packet: %" PRId64, ndatalen);
+
+    if (nwrite <= 0) {
+      switch (nwrite) {
+        case 0:
+          DEBUG(session_, "No data written to packet. Congestion limited?");
+          if (stream_data.id >= 0) ResumeStream(stream_data.id);
+          return congestionLimited(std::move(packet));
+        case NGTCP2_ERR_STREAM_DATA_BLOCKED: {
+          DEBUG(session_,
+                "Stream data is blocked. Congestion limited? We'll try again.");
+          session().StreamDataBlocked(stream_data.id);
+          if (session().max_data_left() == 0) {
+            if (stream_data.id >= 0) ResumeStream(stream_data.id);
+            return congestionLimited(std::move(packet));
+          }
+          CHECK_LE(ndatalen, 0);
+          continue;
+        }
+        case NGTCP2_ERR_STREAM_SHUT_WR: {
+          // Indicates that the writable side of the stream has been closed
+          // locally or the stream is being reset. In either case, we can't send
+          // any stream data!
+          CHECK_GE(stream_data.id, 0);
+          DEBUG_ARGS(
+              session_, "Stream %" PRIi64 " is not writable.", stream_data.id);
+          // We need to notify the stream that the writable side has been closed
+          // and no more outbound data can be sent.
+          CHECK_LE(ndatalen, 0);
+          auto stream = session_->FindStream(stream_data.id);
+          if (stream) stream->EndWritable();
+          continue;
+        }
+        case NGTCP2_ERR_WRITE_MORE: {
+          DEBUG(session_, "Packet is not yet full. Continuing to write more.");
+          CHECK_GT(ndatalen, 0);
+          if (!StreamCommit(&stream_data, ndatalen)) return session_->Close();
+          pos += ndatalen;
+          continue;
+        }
+      }
+
+      DEBUG(session_,
+            "An unexpected error occurred while write packet data. Closing "
+            "session.");
+      session_->last_error_ = QuicError::ForNgtcp2Error(nwrite);
+      return session_->CloseSilently();
+    }
+
+    pos += nwrite;
+    if (ndatalen > 0 && !StreamCommit(&stream_data, ndatalen)) {
+      // Since we are closing the session here, we don't worry about updating
+      // the pkt tx time. The failed StreamCommit should have updated the
+      // last_error_ appropriately.
+      return session_->CloseSilently();
+    }
+
+    if (stream_data.id >= 0 && ndatalen < 0) ResumeStream(stream_data.id);
+
+    packet->Truncate(nwrite);
+    session_->Send(std::move(packet), path);
+
+    pos = nullptr;
+
+    if (++packetSendCount == maxPacketCount) {
```

Comment:
Is this how you plan to deal with the stream blocked  = infinite resends?

I did it a different way in my fork. In my fork I ensured that a stream that returned NGTCP2_ERR_STREAM_DATA_BLOCKED would not be requeued within this call (https://github.com/HalleyAssist/node/commit/a2c6fbe15585c8c9961760e4a0935d29ffae601c)